File: on1ystar/백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    private static int n, m;
    private static int[][] paper;
    private static final int[] dRow = {0, 1, 0};
    private static final int[] dCol = {1, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        paper = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                paper[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int answer = 0;
        boolean[][] visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                visited[i][j] = true;
                answer = Math.max(answer, dfs(i, j, paper[i][j], 0, 1, visited));
                visited[i][j] = false;
            }
        }

        System.out.println(answer);
    }

    private static int dfs(int row, int col, int sum, int max, int cnt, boolean[][] visited) {
        if (cnt == 4) {
            return Math.max(max, sum);
        }

        for (int i = 0; i < 3; i++) {
            int newRow = row + dRow[i];
            int newCol = col + dCol[i];

            if (isValid(newRow, newCol) && !visited[newRow][newCol]) {
                visited[newRow][newCol] = true;

                if (cnt == 2) {
                    max = Math.max(max, dfs(row, col, sum + paper[newRow][newCol], max, cnt + 1, visited));
                }

                max = Math.max(max, dfs(newRow, newCol, sum + paper[newRow][newCol], max, cnt + 1, visited));
                visited[newRow][newCol] = false;
            }
        }

        return max;
    }

    private static boolean isValid(int row, int col) {
         return row >= 0 && row < n && col >= 0 && col < m;
    }
}

File: eello/백준/Gold/14500. 테트로미노/테트로미노.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer
import kotlin.math.max

private enum class Direction(val r: Int, val c: Int) {
    UP(-1, 0),
    RIGHT(0, 1),
    DOWN(1, 0),
    LEFT(0, -1),
    ;
}

private data class Pos(val r: Int, val c: Int) {
    operator fun plus(dir: Direction) = Pos(r + dir.r, c + dir.c)
}

private class Tetronomino(val n: Int, val m: Int, val grid: Array<IntArray>) {

    var answer = 0
        private set

    private val elementMaxValue = grid.maxOf { it.max() }

    fun solve(): Int {
        answer = 0

        val visited = Array(n) { BooleanArray(m) }
        for (r in 0..<n) {
            for (c in 0..<m) {
                visited[r][c] = true
                make(Pos(r, c), grid[r][c], 1, visited)
                visited[r][c] = false
            }
        }

        return answer
    }

    private fun make(cur: Pos, sum: Int, depth: Int, visited: Array<BooleanArray>) {
        if (depth == 4) {
            answer = max(answer, sum)
            return
        }

        if (sum + (4 - depth) * elementMaxValue <= answer) {
            return
        }

        for (dir in Direction.entries) {
            val nxt = cur + dir
            if (isInBounds(nxt) && !visited[nxt.r][nxt.c]) {
                val nxtSum = sum + grid[nxt.r][nxt.c]

                if (depth == 2) {
                    visited[nxt.r][nxt.c] = true
                    make(cur, nxtSum, depth + 1, visited)
                    visited[nxt.r][nxt.c] = false
                }

                visited[nxt.r][nxt.c] = true
                make(nxt, nxtSum, depth + 1, visited)
                visited[nxt.r][nxt.c] = false
            }
        }
    }

    private fun isInBounds(pos: Pos) = pos.r in grid.indices && pos.c in grid[0].indices
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    val grid = Array(n) {
        val line = StringTokenizer(br.readLine())
        IntArray(m) { line.nextToken().toInt() }
    }

    print(Tetronomino(n, m, grid).solve())
}